### DISCUSSION TITLE
real-time collaborative text editing for pair programming

### DISCUSSION THREAD

**Discussion by skyfaller** [helix-editor/helix#3737] [Ideas]
Hi folks, has anyone thought about using Helix for [real-time collaborative text editing](https://en.wikipedia.org/wiki/Collaborative_real-time_editor), e.g. for pair programming? I recognize that the conventional advice for terminal-based text editors is to just [stick it in tmux](https://www.hamvocke.com/blog/remote-pair-programming-with-tmux/), but I was wondering if it would be possible/desirable to design a different workflow, perhaps without the external dependency of tmux.

One feature that stands out to me is that Helix supports multiple cursors. This seems like a crucial feature for pair programming, so perhaps we're already partway there?

UPDATE: Ethersync is a project to enable collaborative editing in an editor-agnostic way, using peer-to-peer networking so you don't need to install a server: https://github.com/ethersync/ethersync

To create a plugin for it, you can build off of an LSP plugin: https://ethersync.github.io/ethersync/editor-plugin-dev-guide.html

helix has LSP built-in, and has no plugin system. Does that mean it would be most appropriate to build ethersync support into helix itself?

**Comment by Omnikar** [comment]
Probably relevant: #312

**Comment by mbrea-c** [comment]
My 2 cents: this would be a killer feature IMO. While not useful 100% of the time, there are many tasks in programming that are very easily parallelizable: writing unit tests, certain kinds of refactoring, any kind of repetitive boilerplate, etc. A friend and I did some hacking sessions using the Zed editor on some project we were working on and we were able to get things done much more smoothly than with traditional _"one person codes and the second watches"_ style of pair programming. Plus, it was more fun, for what that's worth.

By sharing a tmux session all participants are still sharing the same cursor, so I consider it to be closer to traditional pair programming than _real_ collaborative editing.

**Comment by Guekka** [comment]
This might be relevant: [Open Collaboration Tools](https://www.typefox.io/blog/open-collaboration-tools-announcement/)

**Comment by schell** [comment]
This would definitely be nice!

**Comment by danriedl** [comment]
https://github.com/helix-editor/helix/discussions/13226
